### PR TITLE
STARCH-971 base-level event listener and class loader context utilities

### PR DESCRIPTION
**Pull request by mwfyffeiu**

### CODE REVIEW

File: client/src/main/java/edu/iu/client/IuHttp.java [hunk 1]
```diff
@@ -254,28 +256,34 @@ public static <E extends Exception> HttpResponse<InputStream> send(Class<E> exce
 			final var requestHeaders = request.headers();
 			final var requestHeaderMap = requestHeaders.map();
 			if (!requestHeaderMap.isEmpty())
-				// TODO: apply security filter
 				sb.append(' ').append(requestHeaderMap.keySet());
 
+			final var event = new IuHttpClientEvent(uri);
```
[the remark this answers is not in the document]
Comment (mwfyffeiu) [reply client/src/main/java/edu/iu/client/IuHttp.java]: non-issue; uri is effectively final

File: base/src/main/java/edu/iu/IuListener.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	 * Passes an {@link IuObservableEvent} instance from an an upstream resource to
+	 * one or more container-specific event listeners.
+	 * 
+	 * @param event {@link IuObservableEvent}
+	 */
+	static void observe(IuObservableEvent event) {
+		Throwable error = null;
+
+		for (final var listener : ServiceLoader.load(IuListener.class))
+			error = IuException.suppress(error, () -> listener.accept(event));
```
[the remark this answers is not in the document]
Comment (mwfyffeiu) [reply base/src/main/java/edu/iu/IuListener.java]: corrected

File: base/src/main/java/iu/ClassLoaderContext.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	/**
+	 * Gets the context registered for the current thread.
+	 * 
+	 * @return {@link IuClassLoaderContext}
+	 */
+	public static IuClassLoaderContext get() {
+		var loader = Thread.currentThread().getContextClassLoader();
+
+		while (loader != null) {
+			final var context = CONTEXT.get(loader);
```
[the remark this answers is not in the document]
Comment (mwfyffeiu) [reply base/src/main/java/iu/ClassLoaderContext.java]: corrected

**Review by james-anderson-iu — COMMENTED** [review]
## Code Review: STARCH-971 base-level event listener and class loader context utilities

**Modules reviewed:** `base` (primary), `client` (primary), `crypt` (separate)
**Reviewed with:** Copilot

### Summary

This PR introduces a well-structured observable event SPI (`IuObservableEvent`, `IuListener`) using `ServiceLoader`, a ClassLoader context registry (`IuClassLoaderContext`), and HTTP client event emission (`IuHttpClientEvent`). The crypt module receives cleanup and OpenSSL improvements.

### Positive Observations

- **Immutable event pattern** in `IuHttpClientEvent`: `received()` creates a new instance rather than mutating — strong design for multi-listener dispatch.
- **Fail-safe listener dispatch**: Errors in one listener don't prevent delivery to others; full suppression chains via `IuException.suppress()`.
- **ClassLoader parent-chain traversal**: Correctly handles unregistered child classloaders by falling back to registered parents.
- **Good test coverage**: Both public API (`edu.iu.*`) and internal (`iu.*`) layers have dedicated tests covering happy path, error cases, and edge conditions.
- **Crypt cleanup**: Removing `jakarta.annotation` is appropriate — `@Resource`/`@Priority` on internal classes without a container had no runtime effect.

### Prior Reviewer Comments — All Appear Addressed

| Prior feedback | Commit |
|---|---|
| Log level CONFIG → WARNING | `f03568c` |
| ClassLoaderContext synchronization | `3b35e78` |
| ServiceConfigurationError handling | `3b35e78` |
| IuHttp event test conditions | `5903bd4` |
| Bind SPI lookup to base classloader | `b344573` |
| JavaDoc typo | `0c97ac0` |

### Open Questions

1. Is the per-call `ServiceLoader.load()` intentional for dynamic listener hot-swap, or could it be cached?
2. Should `IuObservableEvent` define a `toString()` contract so all event types produce useful log output?
3. Has `default_md = sha-256` been verified against the target OpenSSL version?

File: base/src/main/java/edu/iu/IuListener.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	 * 
+	 * @param event {@link IuObservableEvent}
+	 */
+	static void observe(IuObservableEvent event) {
+		Objects.requireNonNull(event, "event");
+		Throwable error = null;
+
+		Iterator<IuListener> serviceIterator;
+		try {
+			serviceIterator = ServiceLoader.load(IuListener.class, IuListener.class.getClassLoader()).iterator();
```
Comment (james-anderson-iu) [inline base/src/main/java/edu/iu/IuListener.java]: **🟡 IMPORTANT — Performance consideration: ServiceLoader created per invocation**

Every call to `observe()` creates a new `ServiceLoader.load()`, which performs classpath/module scanning for provider declarations. Since this is called on every HTTP request (via `IuHttp`), the overhead accumulates under load.

If the intent is dynamic listener hot-swap, this is acceptable but should be documented. If listeners are stable, caching the `ServiceLoader` (with a `reload()` mechanism) would improve throughput.

**Suggested documentation (if intentional):**
```suggestion
				serviceIterator = ServiceLoader.load(IuListener.class, IuListener.class.getClassLoader()).iterator(); // fresh load per event to support dynamic listener discovery
```

File: base/src/test/java/edu/iu/IuTestListener.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public void beforeEach(ExtensionContext context) throws Exception {
+		final IuListener listener = new IuTestListener();
+		loader = mock(ServiceLoader.class);
+		when(loader.iterator()).thenReturn(IuIterable.iter(listener).iterator());
+		mockSL = mockStatic(ServiceLoader.class);
+		mockSL.when(() -> ServiceLoader.load(IuListener.class, ClassLoader.getSystemClassLoader())).thenReturn(loader);
```
Comment (james-anderson-iu) [inline base/src/test/java/edu/iu/IuTestListener.java]: **🟡 IMPORTANT — Mock classloader doesn't match production code**

Production (`IuListener.java:58`) uses:
```java
ServiceLoader.load(IuListener.class, IuListener.class.getClassLoader())
```

But this mock stubs:
```java
ServiceLoader.load(IuListener.class, ClassLoader.getSystemClassLoader())
```

This works in tests because `IuListener.class.getClassLoader()` resolves to the system classloader in a flat test classpath. But if the module system ever loads `iu.util` from a different classloader, this test would silently stop verifying production behavior.

**Suggested fix:**
```suggestion
		mockSL.when(() -> ServiceLoader.load(IuListener.class, IuListener.class.getClassLoader())).thenReturn(loader);
```

File: client/src/main/java/iu/client/IuHttpClientEvent.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+		return context;
+	}
+
+	@Override
+	public String getAction() {
+		if (responseTime == null)
+			return "send";
+		else
+			return (statusCode < 400 ? "receive " : "error ") + statusCode;
+	}
```
Comment (james-anderson-iu) [inline client/src/main/java/iu/client/IuHttpClientEvent.java:125]: **🟡 IMPORTANT — Missing `toString()` override**

`IuListener.observe()` logs failures with `"event listener failure; " + event`, which calls `toString()`. Without an override, the log output will be the default `iu.client.IuHttpClientEvent@<hashcode>` — unhelpful for diagnosing failures.

**Suggested addition** (after this method):
```java
	@Override
	public String toString() {
		return getType() + "{id=" + id + ", action=" + getAction() + ", uri=" + uri + ", context=" + context + "}";
	}
```

File: crypt/cli/src/main/java/iu/crypt/cli/WebKeyCli.java [hunk 7]
```diff
@@ -581,6 +583,7 @@ private static Path caConfig(Path privateKeyFile, Path databaseFile, Path newCer
 				default_ca = a
 
 				[ a ]
+				default_md = sha-256
```
Comment (james-anderson-iu) [inline crypt/cli/src/main/java/iu/crypt/cli/WebKeyCli.java]: **🟡 IMPORTANT — Potentially non-standard OpenSSL digest name**

The canonical OpenSSL config value is `sha256` (no hyphen). While OpenSSL 3.x may accept `sha-256` via algorithm aliasing, the hyphenated form is non-standard for `.cnf` files and may fail on older OpenSSL versions.

**Suggested fix:**
```suggestion
                            default_md = sha256
```

File: base/src/main/java/edu/iu/IuObservableEvent.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	 * Gets the start time associated with the event, i.e., transaction begin time.
+	 * 
+	 * @return {@link Instant}
+	 */
+	Instant getStartTime();
+
+	/**
+	 * Gets the time the event occurred.
+	 * 
+	 * @return {@link Instant}; null to notify the start of an multi-part
```
Comment (james-anderson-iu) [inline base/src/main/java/edu/iu/IuObservableEvent.java]: **🟢 SUGGESTION — Grammar and completeness**

"an multi-part" → "a multi-part", and the sentence is incomplete.

```suggestion
	 * @return {@link Instant}; null to indicate the start of a multi-part event sequence
```

File: client/src/main/java/edu/iu/client/IuHttp.java [hunk 9]
```diff
@@ -254,28 +256,34 @@ public static <E extends Exception> HttpResponse<InputStream> send(Class<E> exce
...
+			IuListener.observe(event);
+			
 			final HttpResponse<InputStream> response;
 			try {
 				response = HTTP.send(request, BodyHandlers.ofInputStream());
 			} catch (Throwable e) {
 				final var m = "HTTP connection failed " + sb;
 				LOG.log(Level.INFO, e, () -> m);
+				
+				IuListener.observe(event.received(503));
```
Comment (james-anderson-iu) [inline client/src/main/java/edu/iu/client/IuHttp.java]: **🟢 SUGGESTION — Document the synthetic 503 convention**

When `HTTP.send()` throws (DNS failure, TLS error, timeout, etc.), a synthetic 503 is emitted. Downstream listeners have no way to distinguish this from a real server-returned 503. A brief comment would help future maintainers and listener implementors.

```suggestion
					IuListener.observe(event.received(503)); // synthetic 503: no actual HTTP response received
```

File: crypt/cli/src/main/java/iu/crypt/cli/WebKeyCli.java [hunk 10]
```diff
@@ -495,10 +495,12 @@ public static WebKey self(WebKey jwk) {
 		final var keyType = jwk.getType();
 		final var privateKey = Objects.requireNonNull(jwk.getPrivateKey(), "Missing private key");
 		final var privateKeyFile = IuProcess.temp(PemEncoded::print, privateKey);
+		final var cn = jwk.getKeyId().replaceAll("([+=/])", "\\\\$1");
 		final var pemCert = IuProcess.exec( //
 				"openssl", "req", "-x509", "-key", privateKeyFile.toString(), "-days", Integer.toString(days()), //
-				"-subj", subjectOrg() + "/CN=" + jwk.getKeyId().replaceAll("([+=/])", "\\\\$1"), //
+				"-subj", subjectOrg() + "/CN=" + cn, //
 				"-addext", "basicConstraints=CA:false", //
+				"-addext", "subjectAltName=DNS:" + cn + "", //
```
Comment (james-anderson-iu) [inline crypt/cli/src/main/java/iu/crypt/cli/WebKeyCli.java]: **🟢 SUGGESTION — Dead code: trailing empty string concatenation**

The `+ ""` at the end has no effect. Looks like a leftover from a previous edit.

```suggestion
				"-addext", "subjectAltName=DNS:" + cn, //
```

File: base/src/main/java/module-info.java [hunk 11]
```diff
@@ -29,11 +29,15 @@
...
+ * @uses edu.iu.IuListener For observing container-level events
+ */
+module iu.util {
 	exports edu.iu;
-	
-	requires transitive java.logging;
+
+	requires transitive java.logging;
+
+	uses edu.iu.IuListener;
```
Comment (james-anderson-iu) [inline base/src/main/java/module-info.java:42]: **🟢 SUGGESTION — Missing newline at end of file**

The file ends without a trailing newline (confirmed via hex dump: last byte is `7d` = `}`). This causes noisy diffs on future edits and violates POSIX text file conventions. Add a newline after the closing `}`.

**Review by james-anderson-iu — COMMENTED** [review]
**Performance Review** — focused on hot-path overhead introduced by event emission in the `base` and `client` modules.
**Reviewed with:** Copilot

The overall design (immutable event objects, SPI-based listeners, class-loader context tagging) is sound. The findings below address per-request costs that accumulate under concurrency.

File: client/src/main/java/iu/client/IuHttpClientEvent.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+	private final Instant responseTime;
+	private final int statusCode;
+
+	/**
+	 * Constructor.
+	 * 
+	 * @param uri outbound request URI
+	 */
+	public IuHttpClientEvent(URI uri) {
+		this.id = IdGenerator.generateId();
```
Comment (james-anderson-iu) [inline client/src/main/java/iu/client/IuHttpClientEvent.java:59]: **🟡 IMPORTANT — `new SecureRandom()` allocated per event amplifies entropy cost in hot path**

`IdGenerator.generateId()` (line 73 of `IdGenerator.java`) creates a **new `SecureRandom` instance** on every call. This constructor involves native seeding (`CryptGenRandom` on Windows, `/dev/urandom` on Linux). Under burst concurrency, JVM-level entropy pool locks can serialize callers.

This is a pre-existing issue in `IdGenerator`, but this PR places it in the per-HTTP-request critical path where it was not previously invoked. Each outbound HTTP call now pays this cost.

**Suggested mitigation** (in `IdGenerator`, outside this PR’s scope but worth a follow-up):
```java
private static final SecureRandom RANDOM = new SecureRandom();
// then: RANDOM.nextBytes(rawId);
```
Modern `SecureRandom` is thread-safe and auto-reseeds.

File: client/src/main/java/edu/iu/client/IuHttp.java [hunk 13]
```diff
@@ -254,28 +256,34 @@ public static <E extends Exception> HttpResponse<InputStream> send(Class<E> exce
 			final var requestHeaders = request.headers();
 			final var requestHeaderMap = requestHeaders.map();
 			if (!requestHeaderMap.isEmpty())
-				// TODO: apply security filter
 				sb.append(' ').append(requestHeaderMap.keySet());
 
+			final var event = new IuHttpClientEvent(request.uri());
```
Comment (james-anderson-iu) [inline client/src/main/java/edu/iu/client/IuHttp.java:261]: **🟡 IMPORTANT — Event construction + 2× `ServiceLoader.load()` per request even when no listeners exist**

Every HTTP request now executes this sequence unconditionally:
1. `new IuHttpClientEvent(uri)` — `IdGenerator.generateId()` (SecureRandom), `Instant.now()`, `ClassLoaderContext.get()` (synchronized)
2. `IuListener.observe(event)` — fresh `ServiceLoader.load()` + iterator
3. `IuListener.observe(event.received(status))` — second fresh `ServiceLoader.load()` + iterator

When no listeners are registered (the default until a container wires one), this is pure overhead with zero benefit.

**Suggested optimization — fast-path bypass:**
```java
private static volatile boolean listenersChecked;
private static volatile boolean hasListeners;

static void observe(IuObservableEvent event) {
    if (listenersChecked && !hasListeners) return;
    // ... existing logic ...
    if (!listenersChecked) {
        hasListeners = serviceIterator.hasNext();
        listenersChecked = true;
    }
}
```
Alternatively, eagerly resolve providers at module init and store in an immutable list, avoiding `ServiceLoader.load()` per call entirely.

File: base/src/main/java/iu/ClassLoaderContext.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+	 * Gets the context registered for the current thread.
+	 * 
+	 * @return {@link IuClassLoaderContext}
+	 */
+	public static IuClassLoaderContext get() {
+		var loader = Thread.currentThread().getContextClassLoader();
+
+		while (loader != null) {
+			final IuClassLoaderContext context;
+			synchronized (CONTEXT) {
```
Comment (james-anderson-iu) [inline base/src/main/java/iu/ClassLoaderContext.java:86]: **🟢 SUGGESTION — Synchronized lock acquired per ancestor in class-loader walk creates contention under load**

Now that synchronization was correctly added, `get()` acquires `synchronized(CONTEXT)` once per ancestor class loader checked. With application class loaders several levels deep (common in containers), this means 2–3 monitor acquisitions per event.

Combined with the two `IuListener.observe()` calls per HTTP request (each triggers `IuClassLoaderContext.getContext()` via `IuHttpClientEvent` construction), this monitor becomes a serial bottleneck under high concurrency.

**Suggested optimization** — since the thread context class loader rarely changes mid-request, a `ThreadLocal` cache would eliminate repeated lookups:
```java
private static final ThreadLocal<IuClassLoaderContext> CACHE = new ThreadLocal<>();

public static IuClassLoaderContext get() {
    var cached = CACHE.get();
    if (cached != null) return cached;
    // ... existing walk logic ...
    CACHE.set(result);
    return result;
}
```
Alternatively, a `ReadWriteLock` would allow concurrent readers without contention (writes happen only at registration time).

File: client/src/main/java/iu/client/IuHttpClientEvent.java [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+		this.startTime = Instant.now();
+		this.uri = uri;
+		this.context = IuClassLoaderContext.getContext().getName();
```
Comment (james-anderson-iu) [inline client/src/main/java/iu/client/IuHttpClientEvent.java:62]: ### 🟡 IMPORTANT — Information Disclosure: Full Request URI Captured in Event

The constructor stores `uri` verbatim and `getUri()` exposes it to all `IuListener` SPI providers. If the HTTP request URI contains sensitive query parameters (OAuth tokens, API keys, session identifiers, PII), these are broadcast to every service-loaded listener.

**Impact:** A compromised or over-privileged `IuListener` provider could harvest credentials or PII from outbound HTTP calls.

**Recommendation:** Either:
- Sanitize the URI before storing (strip query params, or redact known-sensitive parameter values), or
- Document the trust model explicitly in the `IuListener.observe()` Javadoc: only container-provisioned, fully-trusted listener implementations should appear on the base module's classpath.

Given that module boundaries already restrict who can provide an `IuListener`, documenting the trust assumption may be sufficient.

File: crypt/cli/src/main/java/iu/crypt/cli/WebKeyCli.java [hunk 16]
```diff
@@ -810,6 +815,8 @@ public static void main(String[] arg) {
 				} else if (cmd.equals("export")) {
 					if (inputKey != null)
 						export(System.out, inputKey);
+					else if (arg.length == 1)
+						export(System.out, inputCa.getJwk());
```
Comment (james-anderson-iu) [inline crypt/cli/src/main/java/iu/crypt/cli/WebKeyCli.java]: ### 🟡 IMPORTANT — CA Private Key Export Without Safety Rail

This new code path outputs the CA's private key + certificate to stdout when no serial argument is given. This makes it easy to accidentally exfiltrate the CA signing key by running `echo <ca-json> | cli export` with no extra argument (distinguishing "export the CA itself" from "export a cert by serial" only by argument count).

**Impact:** In a pipeline or automated context, the CA private key could end up in logs, artifacts, or sent to an unintended recipient.

**Recommendation:** Consider requiring an explicit flag (e.g., `export --ca`) or adding a stderr warning when outputting CA private material, to prevent accidental disclosure. The test confirms this is intentional behavior, but the UX lacks a guardrail.

File: base/src/main/java/edu/iu/IuListener.java (hunk 4 above)
Comment (james-anderson-iu) [inline base/src/main/java/edu/iu/IuListener.java]: ### 🟢 SUGGESTION — Document ServiceLoader Trust Boundary

Binding the ServiceLoader lookup to `IuListener.class.getClassLoader()` (rather than the thread context classloader) is a deliberate and correct security-scoping decision — it ensures only listeners visible to the `iu.util` module are invoked.

**Recommendation:** Add a brief Javadoc note on `observe()` documenting this trust model for future maintainers:

```suggestion
			serviceIterator = ServiceLoader.load(IuListener.class, IuListener.class.getClassLoader()).iterator();
			// Bound to base module classloader: only container-provisioned listeners are invoked
```

File: base/src/main/java/iu/ClassLoaderContext.java [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+		CONTEXT = context;
+	}
+
+	/**
+	 * Registers a context with a {@link ClassLoader}
+	 * 
+	 * @param context {@link IuClassLoaderContext}
+	 * @param loader  {@link ClassLoader}
+	 */
+	public static void register(IuClassLoaderContext context, ClassLoader loader) {
```
Comment (james-anderson-iu) [inline base/src/main/java/iu/ClassLoaderContext.java:65]: ### 🟢 SUGGESTION — Registration Authorization Note

`register()` accepts any caller that can provide a non-null ClassLoader reference. The security boundary here is the module system itself: the `iu` package is not exported, so only code within `iu.util` or code that routes through the public `IuClassLoaderContext.register()` facade can reach this. The `assertNotOpen()` guard in the static initializer adds a runtime check that the module isn't reflectively opened.

This is adequate. Consider adding a brief note to the `IuClassLoaderContext.register()` Javadoc stating it should only be invoked by container initialization code.

File: crypt/cli/src/main/java/iu/crypt/cli/WebKeyCli.java (hunk 10 above)
Comment (james-anderson-iu) [inline crypt/cli/src/main/java/iu/crypt/cli/WebKeyCli.java]: ### 🟢 SUGGESTION — SAN DNS Value May Contain Non-DNS Characters

The `cn` variable (key ID with `+=/` backslash-escaped) is used directly as a DNS SAN. While not a vulnerability (the value originates from a trusted `IdGenerator`, not user input), DNS SANs with backslash-escaped characters may cause TLS validation issues in some clients that perform strict hostname matching.

**Recommendation:** Verify downstream TLS clients accept these SAN values. If the key ID contains non-DNS-safe characters, consider using a sanitized or hashed form for the SAN while keeping the original for the CN.

File: crypt/cli/src/main/java/iu/crypt/cli/WebKeyCli.java (hunk 7 above)
Comment (james-anderson-iu) [inline crypt/cli/src/main/java/iu/crypt/cli/WebKeyCli.java]: ### 🟢 POSITIVE — Explicit `default_md` is a Security Improvement

Adding `default_md = sha-256` prevents OpenSSL from falling back to a potentially weaker default digest algorithm (varies by OpenSSL version/config). Good hardening change.

**Review by james-anderson-iu — COMMENTED** [review]
## Security Review — `base`, `client`, and `crypt` modules

**Reviewed with:** Copilot

Overall this is well-structured work. The module boundary enforcement via `IuObject.assertNotOpen()`, the immutable event design, and the deliberate binding of ServiceLoader to the base module's classloader are all sound security decisions.

The main security-relevant items are below, ordered by severity. All prior reviewer conversations appear to have been addressed by subsequent commits (synchronization, logging level, classloader binding, test mocking).

File: base/src/main/java/edu/iu/IuListener.java (hunk 4 above)
Comment (mwfyffeiu) [reply base/src/main/java/edu/iu/IuListener.java]: Updated to use a cached ServiceLoader; hot reload() is not important. Listener configuration is part of base classpath.

File: client/src/main/java/edu/iu/client/IuHttp.java (hunk 9 above)
Comment (mwfyffeiu) [reply client/src/main/java/edu/iu/client/IuHttp.java]: Switched to use 0 instead of 503

File: base/src/main/java/iu/ClassLoaderContext.java (hunk 14 above)
Comment (mwfyffeiu) [reply base/src/main/java/iu/ClassLoaderContext.java:86]: Contention is allowable here; expecting o(100) or less locking for a 0-cost get. Not a cache, so ThreadLocal doesn't make sense.

File: client/src/main/java/iu/client/IuHttpClientEvent.java (hunk 15 above)
Comment (mwfyffeiu) [reply client/src/main/java/iu/client/IuHttpClientEvent.java:62]: Documented the trust model explicitly

File: crypt/cli/src/main/java/iu/crypt/cli/WebKeyCli.java (hunk 16 above)
Comment (mwfyffeiu) [reply crypt/cli/src/main/java/iu/crypt/cli/WebKeyCli.java]: added wellKnown() to CA JWK export to omit private keys.

File: client/src/main/java/edu/iu/client/IuHttp.java (hunk 13 above)
Comment (mwfyffeiu) [reply client/src/main/java/edu/iu/client/IuHttp.java:261]: ServiceLoader.load() is now cached, so not an issue here. IdGenerator SecureRandom is also cached now. synch lock on ClassLoaderContext is allowable

**Review by james-anderson-iu — APPROVED** [review]
